frontend/apps/reader/modules/readerback.lua

```diff
@@ -1,89 +1,193 @@
+            -- objects ReaderPaging/ReaderView may still modify
+            local res = {}
+            for i=1, #current_location do
+                res[i] = util.tableDeepCopy(current_location[i])
+            end
+            return res
+        end
     else
-        current_location = {
+        return {
             xpointer = self.ui.rolling:getBookLocation(),
         }
     end
-
-    return current_location
 end
 
-local ignore_location
+function ReaderBack:_areLocationsSimilar(location1, location2)
+    if self.ui.document.info.has_pages then
+        -- locations are arrays of k/v tables
+        if #location1 ~= #location2 then
+            return false
+        end
+        for i=1, #location1 do
+            if not util.tableEquals(location1[i], location2[i]) then
+                return false
+            end
+        end
+        return true
+    else
+        return location1.xpointer == location2.xpointer
+    end
+end
 
-function ReaderBack:addCurrentLocationToStack()
-    local location_stack = self.location_stack
+function ReaderBack:_addPreviousLocationToStack()
     local new_location = self:_getCurrentLocation()
 
-    if util.tableEquals(ignore_location, new_location) then return end
-
-    table.insert(location_stack, new_location)
-
-    if #location_stack > self.max_stack then
-        table.remove(location_stack, 1)
+    if self.cur_location and new_location then
+        if self:_areLocationsSimilar(self.cur_location, new_location) then
+            -- Unchanged, don't add it yet
+            return
+        end
+        table.insert(self.location_stack, self.cur_location)
+        if #self.location_stack > self.max_stack then
+            table.remove(self.location_stack, 1)
+        end
     end
-end
 
--- Scroll mode crengine
-function ReaderBack:onPosUpdate()
-    if self.disabled then return end
-    self:addCurrentLocationToStack()
-end
-
--- Paged media
-function ReaderBack:onPageUpdate()
-    if self.disabled then return end
-    self:addCurrentLocationToStack()
+    if new_location then
+        self.cur_location = new_location
+    end
 end
 
 -- Called when loading new document
 function ReaderBack:onReadSettings(config)
     self.location_stack = {}
+    self.cur_location = nil
 end
 
-function ReaderBack:onBack()
-    local location_stack = self.location_stack
+function ReaderBack:_onViewPossiblyUpdated()
+    if G_reader_settings:readSetting("back_in_reader") == "previous_read_page" then
+        -- As multiple modules will have their :onPageUpdate()/... called,
+        -- and some of them will set up the new page with it, we need to
+        -- delay our handling after all of them are called (otherwise,
+        -- depending on the order of the calls, we may be have the location
+        -- of either the previous page or the current one).
+        UIManager:nextTick(self._addPreviousLocationToStackCallback)
+    end
+    self.back_resist = nil
```

Comment:
I'm not super-fond of the variable name, but I don't really have any better idea either ;D.